Initial version

rgaudin opened the pull request:
Initial version of the configuration scripts collection:

- `offspot-config-from-file`: expected to be ran on boot via a systemd service. Launches the other scripts from sub-parts of the passed YAML file.
- `offspot-config-timezone`: sets the timezone
- `offspot-config-hostname`: sets local hostname
- `offspot-config-ethernet`: configures ethernet interface: dhcp or static
- `offspot-config-ap`: configures the wireless Access Point.
- `offspot-config-containers`: configures (writes) the docker-compose file

Behavior is better describe in README.

## Testing

Follow install instructions in README or `python setup.py bdist_upload` and copy it to pi. Then:

```sh
apt install hostapd dnsmasq dhcpcd5 python3-yaml
systemctl unmask hostapd
systemctl disable hostapd dnsmasq
pip3 install offspot_runtime_config-1.0-py3-none-any.whl
```

```sh
offspot-config-fromfile --debug /boot/offspot.yaml
```

#### Samples

**Full**

```yaml
---
timezone: Africa/Bamako
hostname: renaud
ethernet:
  type: dhcp
ap:
  ssid: Master π
  passphrase: 3.141592
  as-gateway: true
containers:
  services:
    httpd:
      container_name: base-httpd
      image: "ghcr.io/offspot/base-httpd:1.0"
      pull_policy: never
      read_only: true
      restart: unless-stopped
      ports:
        - "80:80"
      volumes:
        - type: bind
          source: /boot/offspot.json
          target: /var/www/offspot.json
          read_only: true
        - type: bind
          source: /boot/offspot.yaml
          target: /var/www/offspot.yaml
          read_only: true
```

**Static network**

```yaml
ethernet:
  type: static
  address: 10.0.0.1
  routers:
  - 10.0.0.10
  dns:
  - 10.0.0.10
  - 8.8.8.8
```

**AP scenario A: Hotspot-only**

```yaml
ap:
  ssid: Welcome WiFi
```

**AP scenario B: internet relay (sharing ethernet connexion along offspot)**

```yaml
ap:
  ssid: Welcome WiFi
  as-gateway: true
```

**AP scenario C: WiFi+wired hotspot**

you'll need to configure static ethernet for this one

```yaml
ap:
  ssid: Welcome WiFi
  other-interfaces:
  - eth0
ethernet:
  type: static
  address: 10.0.0.1
  routers:
  - 10.0.0.1
  dns:
  - 10.0.0.1
```

**AP scenario D: Hotspot + existing LAN**

Offspot available via WiFi and on existing LAN

```yaml
ap:
  ssid: Welcome WiFi
  nodhcp-interfaces:
  - eth0
```

kelson42 commented:
@rgaudin There is nothing about CI/CD, is that normal?

rgaudin commented:
> @rgaudin There is nothing about CI/CD, is that normal?

Yes. PR is long enough ; will add QA and release once this is reviewed. Kinda depends on feedback (do we want to keep it here and publish ? Is the package name OK?).

Note that I don't plan on adding automated tests because it's not really testable.

Bastien-BSF commented on README.md:
````diff
@@ -1,2 +1,196 @@
+offspot-config-fromfile --debug /boot/offspot.yaml
+```
+
+- `--debug` will show you what exact parameters where passed to individual scripts so you can manually launch them should there be an issue.
````
```suggestion
- `--debug` will show you what exact parameters were passed to individual scripts so you can manually launch them should there be an issue.
```

Bastien-BSF commented on README.md:
```diff
@@ -1,2 +1,196 @@
+| `hide`       | `boolean`  | no | Hide SSID (Clients must know and enter its name to connect)                                                      |
+| `interface`  | `string`   | no | Interface to configure AP for. Defaults to `wlan0`                                                               |
+| `dhcp-range` | `string`   | no | IP range for AP clients. `start,end,subnet,ttl` format. Default: `.100-.240` from address                        |
+| `network`    | `[string]` | no | Network to advertise DHCP on. Defaults to `.0/24` from address                                                   |
```
```suggestion
| `network`    | `[string]` | no | Network to advertise DHCP on. Defaults to `.0/24` from address                                                   |
| `nodhcp-interfaces` | `[string]` | no | Interfaces where the DHCP server will not run |
```

Bastien-BSF commented on README.md:
````diff
@@ -1,2 +1,197 @@
+**⚠️ Warning**: only tested on offspot base-image (raspiOS bullseye)
+
+```sh
+apt install hostapd dnsmasq dhcpcd5 python3-yaml
````
```suggestion
apt install hostapd dnsmasq dhcpcd5 python3-yaml python3-pip
```

Bastien-BSF commented:
I have tried something that is probably not meant to be supported:
I have installed `base-image` while supplying `WPA-ESSID=` in order to connect it to my local wifi network.
Then I have installed `runtime-config` and added the `/boot/offspot.yaml` with the following configuration:
```
---
ap:
  ssid: Welcome WiFi
```
After booting, I could not see the new hotspot. But the system did connect properly to the local wifi network and `wlan0` had IPs from the wifi network dhcp and also the default `192.168.2.1`.

When I execute `offspot-config-fromfile --debug /boot/offspot.yaml` manually I get the following output:

```
fromfile INFO: Starting offspot-runtime-config off /boot/offspot.yaml
fromfile DEBUG: [ap] config change requested
fromfile DEBUG: command=['/usr/bin/env', 'offspot-config-ap', '--debug', 'Welcome WiFi']
ap INFO: Configuring WiFi AP
ap DEBUG: command=['/usr/sbin/rfkill', 'unblock', 'wifi']
ap DEBUG: rfkill status:
ID TYPE      DEVICE TYPE-DESC         SOFT      HARD
 0 wlan      phy0   Wireless LAN unblocked unblocked
 1 bluetooth hci0   Bluetooth    unblocked unblocked

---
ap DEBUG: wireless unblocked
ap DEBUG: hostapd.conf checked and written
ap DEBUG: command=['/usr/bin/systemctl', 'status', 'hostapd']
● hostapd.service - Access point and authentication server for Wi-Fi and Ethernet
     Loaded: loaded (/lib/systemd/system/hostapd.service; disabled; vendor preset: enabled)
     Active: inactive (dead)
       Docs: man:hostapd(8)
ap ERROR: ['/usr/bin/systemctl', 'status', 'hostapd'] failed with returncode 3
ap DEBUG: command=['/usr/bin/systemctl', 'start', 'hostapd']
ap DEBUG: hostapd restarted
ap DEBUG: command=['/usr/sbin/ifconfig', 'wlan0', '192.168.2.1', 'netmask', '255.255.255.0', 'up']
ap DEBUG: ip-address set
ap DEBUG: command=['/usr/sbin/dnsmasq', '--test', '-C', '/tmp/tmpewepn7yt.conf']
dnsmasq: syntax check OK.
ap DEBUG: dnsmasq.conf written
ap DEBUG: command=['/usr/bin/systemctl', 'status', 'dnsmasq']
● dnsmasq.service - dnsmasq - A lightweight DHCP and caching DNS server
     Loaded: loaded (/lib/systemd/system/dnsmasq.service; disabled; vendor preset: enabled)
     Active: inactive (dead)
ap ERROR: ['/usr/bin/systemctl', 'status', 'dnsmasq'] failed with returncode 3
ap DEBUG: command=['/usr/bin/systemctl', 'start', 'dnsmasq']
ap DEBUG: dnsmasq restarted
ap DEBUG: command=['/usr/sbin/sysctl', '-p']
ap DEBUG: routing enabled
ap DEBUG: masquerade disabling requested
ap DEBUG: forwarding disabling requested
ap INFO: Wireless AP configured
fromfile INFO: [ap] configuration applied
fromfile INFO: runtime-config applied successfuly
```

rgaudin commented:
> I have installed `base-image` while supplying `WPA-ESSID=` in order to connect it to my local wifi network.

How did you do that exactly? the builder script doesn't support those and rewrites the config file completely.

Bastien-BSF commented:
> How did you do that exactly?
If I remember correctly I added those temporarily in the dataclass for this specific build.

rgaudin commented:
@Bastien-BSF just push a couple tiny commits:
- fixed how `fromfile` behaves with malformed YAML: was not exiting directly thus raising an error.
- added `--no-pager` to some `systemctl` commands. Noticed the pager when doing some manual tests. I don't think it was an issue when launched automatically but it's better this way.

rgaudin commented:
> If I remember correctly I added those temporarily in the dataclass for this specific build.

Thank you. It turns out to be harmless as it just triggers wpa_supplicant conf but it is not used as `wlan0` is referenced (by us) in `/etc/network/interfaces` without mentioning wpa_conf.

Bastien-BSF reviewed (approved):
LGTM
